docs: Document cluster types and important namespaces for provider/service development (#65)
openmcp-project/docs: docs/developers/clusters-and-namespaces.md

---
sidebar_position: 1
---

# Clusters and Namespaces

OpenControlPlane spans multiple Kubernetes clusters, each with its own namespace conventions.

## Cluster Overview

OpenControlPlane uses four types of clusters, each with a distinct role:

| Cluster | Purpose | Primary Audience |
|---------|---------|-----------------|
| **Onboarding** | User-facing API surface for managing Projects, Workspaces, and ControlPlanes | End users |
| **Platform** | Runs all operators (openmcp-operator, service providers, cluster providers) and manages per-tenant resources | Platform operators, service provider developers |

> | **MCP** | Per-tenant lightweight Kubernetes cluster that hosts the DomainServiceAPI | End users (via DomainServiceAPI) |

Reviewer: I think at this point in this context it is not clear what DomainServiceAPI means.

```suggestion
| **MCP** | Per-tenant lightweight Kubernetes cluster | End users |
```

> | **Workload** | Per-tenant cluster for running DomainService controllers (optional) | Service provider developers |

Reviewer: Workload clusters are shared by multiple tenants

## Namespace Model

The following diagram shows how namespaces are organized across all four cluster types:

```mermaid
---
config:
  themeVariables:
    fontSize: 25px
---
graph LR
    subgraph OnboardingCluster["Onboarding"]
        ON1["<b>project-platform-team</b><br/>Workspaces, ServiceProviderAPIs"]
        ON2["<b>project-platform-team--ws-dev</b><br/>ControlPlane resources"]
    end

    subgraph PlatformCluster["Platform"]
        PN1["<b>openmcp-system</b><br/>All provider pods, operator"]
        PN2["<b>mcp--&lt;uuid&gt;</b><br/>AccessRequests, secrets for Tenant A"]
        PN3["<b>mcp--&lt;uuid&gt;</b><br/>AccessRequests, secrets for Tenant B"]
    end

    subgraph MCPCluster["MCP (per tenant)"]

>         MN1["<b>crossplane-system</b><br/>Crossplane components"]

Reviewer: For consistency and to prevent confusion about any MCP having crossplane, landscaper and flux, I would prefer the `service-specific namespace` like in the workload cluster box.

        MN2["<b>ls-system-&lt;id&gt;</b><br/>Landscaper components"]
        MN3["<b>flux-system</b><br/>Flux components"]
    end

    subgraph WorkloadCluster["Workload (optional)"]
        WN1["<b>service-specific namespace</b><br/>DomainService controllers"]
    end

    OnboardingCluster ~~~ PlatformCluster
    MCPCluster ~~~ WorkloadCluster

    style OnboardingCluster fill:#f0f9ff,stroke:#0284c7
    style PlatformCluster fill:#fef3c7,stroke:#d97706
    style MCPCluster fill:#f0fdf4,stroke:#16a34a
    style WorkloadCluster fill:#fdf2f8,stroke:#db2777
```

### Onboarding Cluster

The Onboarding Cluster is the entry point for end users. Namespaces follow a hierarchical pattern:

| Namespace | Created By | Contains |
|-----------|-----------|----------|
| `project-<project-name>` | Creating a `Project` | Workspaces, project-scoped resources |
| `project-<project-name>--ws-<workspace-name>` | Creating a `Workspace` | `ManagedControlPlaneV2` resources, ServiceProviderAPI instances |

For example, a project named `platform-team` with a workspace `dev` results in:
- `project-platform-team` for the project
- `project-platform-team--ws-dev` for the workspace

> ServiceProviderAPI CRDs are installed on the Onboarding Cluster so that all tenants can discover available services.

Reviewer: Similar to DomainServiceAPI reference earlier, I think it is not clear what a ServiceProviderAPI is in this context.

```suggestion
[Service Providers](../users/concepts/service-provider.md) install CRDs on the Onboarding Cluster so that all tenants can discover available services.
```

### Platform Cluster

> The Platform Cluster runs all operators and manages per-tenant resources:

Reviewer: ```suggestion
The Platform Cluster runs all operators. Tenant resources are isolated using namespaces:
```

| Namespace | Purpose |
|-----------|---------|
| `openmcp-system` | System namespace where all provider pods (service providers, cluster providers, platform services) and the openmcp-operator run |
| `mcp--<uuid>` | One per MCP tenant. Auto-generated using a deterministic SHAKE128 hash of the MCP name and namespace. Contains AccessRequests, ClusterRequests, and kubeconfig secrets scoped to that tenant |

> The `mcp--<uuid>` namespaces are created and managed automatically by the platform. The [service-provider-runtime](./serviceprovider/02-service-providers.mdx) resolves the correct tenant namespace via the [`ClusterContext`](./serviceprovider/02-service-providers.mdx#createorupdate-operation).

Reviewer: Since this is a general entry point to get an overview, the sp-runtime and cluster context reference is very specific and not necessary at this point.

```suggestion
The `mcp--<uuid>` namespaces are created and managed automatically by the platform.
```

:::info
The `POD_NAMESPACE` environment variable, available to all provider pods, refers to the provider's namespace on the Platform Cluster (typically `openmcp-system`). See the [deployment guide](./serviceprovider/01-deployment.mdx) for all available environment variables.
:::

### MCP Cluster

Each tenant gets a dedicated MCP (Managed Control Plane) Cluster — a lightweight Kubernetes cluster with its own API server and data store. This provides the strongest isolation boundary between tenants.

> | Namespace | Purpose |

Reviewer: Similar as before, I would suggest to use the same table as for workload clusters.

|-----------|---------|
| `crossplane-system` | Crossplane components (installed by service-provider-crossplane) |
| `ls-system-<instance-id>` | Landscaper components (installed by service-provider-landscaper) |
| `flux-system` | Flux components |
| `cert-manager` | cert-manager components |
| `external-secrets` | External Secrets Operator |

Each service provider chooses its own namespace on the MCP cluster — there is no single default namespace convention. For example, Crossplane uses the fixed namespace `crossplane-system`, while Landscaper derives an instance-specific namespace like `ls-system-<id>`. When building a new service provider, you decide which namespace to deploy your resources into.

### Workload Cluster

> Workload Clusters are optional per-tenant clusters provisioned on demand when a service provider requests them. They are used for running DomainService controllers outside the MCP cluster.

Reviewer: Workload clusters are shared multi-tenant clusters. Tenant separation using namespaces should be pointed out as a requirement similar to the platform cluster. Service provider Landscaper is a good example (sp-velero uses the same logic). You can use your the instance-id description from the real world examples section.

| Namespace | Purpose |
|-----------|---------|
| Service-specific (e.g., `envoy-gateway-system`, `velero`) | Where DomainService controllers and their resources run |

> Each MCP that requests a workload cluster gets its own dedicated cluster, maintaining per-tenant isolation. Service providers choose their own namespace naming conventions on the workload cluster.

Reviewer: MCP should be service provider. I suggest to just rephrase this based on my previous comment.
